### PR TITLE
Action._customArguments: Map<String, String>

### PR DESCRIPTION
**Pull request by jmfayard**
I've added `Action._customArguments: Map<String, String>`

### CODE REVIEW

File: wrapper-generator/src/test/kotlin/it/krzeminski/githubactions/wrappergenerator/generation/GenerationTest.kt [hunk 1]
```diff
@@ -44,44 +44,51 @@ class GenerationTest : FunSpec({
         // then
         wrapper shouldBe Wrapper(
             kotlinCode = """
-                // This file was generated using 'wrapper-generator' module. Don't change it by hand, your changes will
```
Comment (jmfayard) [inline wrapper-generator/src/test/kotlin/it/krzeminski/githubactions/wrappergenerator/generation/GenerationTest.kt:47]: I find super painful to have to cpopy/paste and then indent and escape the content of the file inside `GenerationTest.kt`

What I usually do instead is: 

- the current content of `DeprecatedActionV2.kt` is used as the expected value for the test
- the script writes the new content to `DeprecatedActionV2Actual.kt`
- if the content don't match, an assert is thrown and we simply have to compare `DeprecatedActionV2.kt` with `DeprecatedActionV2Actual.kt`, and update the changes that look good
- if the content do match, the test deletes the file `DeprecatedActionV2Actual.kt`

Comment (krzema12) [reply wrapper-generator/src/test/kotlin/it/krzeminski/githubactions/wrappergenerator/generation/GenerationTest.kt:47]: Makes sense to me, we wouldn't have the same file in two places. Would you like to make such change in a separate PR, in existing tests?

**Comment by krzema12** [comment]
> I wonder if I should add a way to override the default version, as suggested in https://github.com/krzema12/github-actions-kotlin-dsl/issues/74

Would be cool (maybe `_overrideVersion`?), but it also deserves a separate PR :pray:

File: library/src/test/kotlin/it/krzeminski/githubactions/yaml/StepsToYamlTest.kt [hunk 2]
```diff
@@ -345,4 +345,34 @@ class StepsToYamlTest : DescribeSpec({
                              |    compiler: latexmk""".trimMargin()
         }
     }
+
+    it("renders with action with custom arguments") {
```
Comment (krzema12) [inline library/src/test/kotlin/it/krzeminski/githubactions/yaml/StepsToYamlTest.kt:349]: The test itself looks good, but should be nested within `describe("external action step") {`.

**Review by krzema12 — CHANGES_REQUESTED** [review]

File: wrapper-generator/src/test/kotlin/it/krzeminski/githubactions/wrappergenerator/generation/GenerationTest.kt [hunk 3]
```diff
@@ -126,55 +133,62 @@ class GenerationTest : FunSpec({
...
```
Comment (krzema12) [inline wrapper-generator/src/test/kotlin/it/krzeminski/githubactions/wrappergenerator/generation/GenerationTest.kt:129]: I see the whole contents of the expected files replaced, perhaps some whitespace changes? Let's make the refactoring you suggested first (having expected contents only in separate files), then this problem will disappear :magic

File: wrapper-generator/src/test/kotlin/it/krzeminski/githubactions/wrappergenerator/generation/GenerationTest.kt (hunk 1 above)
Comment (jmfayard) [reply wrapper-generator/src/test/kotlin/it/krzeminski/githubactions/wrappergenerator/generation/GenerationTest.kt:47]: https://github.com/krzema12/github-actions-kotlin-dsl/pull/140

File: wrapper-generator/src/test/kotlin/it/krzeminski/githubactions/wrappergenerator/generation/GenerationTest.kt (hunk 3 above)
Comment (jmfayard) [reply wrapper-generator/src/test/kotlin/it/krzeminski/githubactions/wrappergenerator/generation/GenerationTest.kt]: done

**Comment by jmfayard** [comment]
@krzema12 I'm done, I integrated the changes from the main branch and rebased upon it

File: wrapper-generator/src/main/kotlin/it/krzeminski/githubactions/wrappergenerator/generation/Generation.kt [hunk 4]
```diff
@@ -262,7 +277,12 @@ private fun Metadata.primaryConstructor(inputTypings: Map<String, Typing>, coord
                     .defaultValueIfNullable(input)
                     .addKdoc(input.description)
                     .build()
-            }
+            }.plus(
+                ParameterSpec.builder(CUSTOM_ARGUMENTS, Types.MapStringString)
+                    .defaultValue("mapOf()")
+                    .addKdoc("Type-unsafe map where you can put any inputs that is not yet supported by the wrapper")
```
Comment (krzema12) [inline wrapper-generator/src/main/kotlin/it/krzeminski/githubactions/wrappergenerator/generation/Generation.kt:283]:
```suggestion
                    .addKdoc("Type-unsafe map where you can put any inputs that are not yet supported by the wrapper")
```

File: library/src/gen/kotlin/it/krzeminski/githubactions/actions/googlegithubactions/SetupGcloudV0.kt [hunk 5]
```diff
@@ -81,74 +86,76 @@ public class SetupGcloudV0(
...
 
-        public object Cbt : SetupGcloudV0.Component("cbt")
-
-        public object Bigtable : SetupGcloudV0.Component("bigtable")
-
-        public object Datalab : SetupGcloudV0.Component("datalab")
+        public object AppEngineGo : SetupGcloudV0.Component("app-engine-go")
 
-        public object CloudDatastoreEmulator : SetupGcloudV0.Component("cloud-datastore-emulator")
+        public object AppEngineJava : SetupGcloudV0.Component("app-engine-java")
```
Comment (krzema12) [inline library/src/gen/kotlin/it/krzeminski/githubactions/actions/googlegithubactions/SetupGcloudV0.kt:104]: It's weird these changes appear here. Could you rebase on newest master again please?

File: library/src/gen/kotlin/it/krzeminski/githubactions/actions/azure/DockerLoginV1.kt [hunk 6]
```diff
@@ -35,5 +40,7 @@ public class DockerLoginV1(
             password?.let { "password" to it },
             loginServer?.let { "login-server" to it },
         ).toTypedArray()
-    )
+    ).also {
```
Comment (krzema12) [inline library/src/gen/kotlin/it/krzeminski/githubactions/actions/azure/DockerLoginV1.kt:43]: Just got me thinking: why not simply
```kotlin
) + _customArguments
```
? Less imperative, more functional approach.

File: library/src/gen/kotlin/it/krzeminski/githubactions/actions/azure/DockerLoginV1.kt [hunk 7]
```diff
@@ -26,7 +27,11 @@ public class DockerLoginV1(
     /**
      * Container registry server url
      */
-    public val loginServer: String? = null
+    public val loginServer: String? = null,
+    /**
+     * Type-unsafe map where you can put any inputs that is not yet supported by the wrapper
+     */
+    public val _customArguments: Map<String, String> = mapOf()
```
Comment (krzema12) [inline library/src/gen/kotlin/it/krzeminski/githubactions/actions/azure/DockerLoginV1.kt:34]: I'm wondering if it should be a LinkedHashMap, consistently with what `toYamlArguments` returns. The reason for it is to have better control over the order of these in the resulting YAML, at the cost of having to write an extra `linked` in `linkedMapOf(...)`.

**Review by krzema12 — CHANGES_REQUESTED** [review]

File: library/src/gen/kotlin/it/krzeminski/githubactions/actions/azure/DockerLoginV1.kt (hunk 6 above)
Comment (jmfayard) [reply library/src/gen/kotlin/it/krzeminski/githubactions/actions/azure/DockerLoginV1.kt:43]: That would be of type `Map`, necessitating the whole expression to be wrapped a third time inside a `LinkedHashMap(...)`

File: library/src/gen/kotlin/it/krzeminski/githubactions/actions/azure/DockerLoginV1.kt (hunk 7 above)
Comment (jmfayard) [reply library/src/gen/kotlin/it/krzeminski/githubactions/actions/azure/DockerLoginV1.kt:34]: Actually `mapOf(...)` preserves the order exactly like `linkedMapOf(...)` plus it's inmutable contrary to the later

File: library/src/gen/kotlin/it/krzeminski/githubactions/actions/azure/DockerLoginV1.kt (hunk 6 above)
Comment (jmfayard) [reply library/src/gen/kotlin/it/krzeminski/githubactions/actions/azure/DockerLoginV1.kt]: I have found a better solution

```kotlin
listOfNotNull(
   "foo-bar" to fooBar,
  *_customArguments.toList().toTypedArray(),
).toTypedArray()
```

File: library/src/gen/kotlin/it/krzeminski/githubactions/actions/googlegithubactions/SetupGcloudV0.kt (hunk 5 above)
Comment (jmfayard) [reply library/src/gen/kotlin/it/krzeminski/githubactions/actions/googlegithubactions/SetupGcloudV0.kt:105]: I tried but it's my version that is correct:

```kotlin
EnumTyping("Component",
  listOf( "alpha","anthos-auth", "appctl",)
  ```

Action__customArguments__Map_String__String__by_jmfayard_·_Pull_Request__137_·_krzema12_github-actions-kotlin-dsl

Looks like the wrappers were not re-generated before committing

Comment (krzema12) [reply library/src/gen/kotlin/it/krzeminski/githubactions/actions/googlegithubactions/SetupGcloudV0.kt:105]: Oops, right. I fixed `main` branch added an extra check (https://github.com/krzema12/github-actions-kotlin-dsl/commit/c19510340d91ac070902583146916b872ff17cf9) so it shouldn't happen again.

**Comment by krzema12** [comment]
Looks good, but there's some noise related to cherry-picked changes. Could you please rebase on newest `main`?

**Comment by jmfayard** [comment]
Done

File: library/src/gen/kotlin/it/krzeminski/githubactions/actions/10up/ActionWordpressPluginDeployV2.kt [hunk 8]
```diff
@@ -19,7 +22,11 @@ public class ActionWordpressPluginDeployV2(
     /**
      * Generate package zip file?
      */
-    public val generateZip: Boolean? = null
+    public val generateZip: Boolean? = null,
+    /**
+     * Type-unsafe map where you can put any inputs that are not yet supported by the wrapper
+     */
+    public val _customArguments: Map<String, String> = mapOf()
```
Comment (krzema12) [inline library/src/gen/kotlin/it/krzeminski/githubactions/actions/10up/ActionWordpressPluginDeployV2.kt:29]: Let's call it `_customInputs`, to stick to GitHub's nomenclature:
https://github.com/actions/checkout/blob/2d1c1198e79c30cca5c3957b1e3b65ce95b5356e/action.yml#L3

**Review by krzema12 — CHANGES_REQUESTED** [review]
Just this one thing left, and we're merging :rocket:

**Comment by jmfayard** [comment]
Done

**Review by krzema12 — APPROVED** [review]
Thank you! :bow:

**Pull request merged**